src/utils/Fetcher.ts

```diff
@@ -107,54 +119,51 @@ export async function queryGraph<T = any>(
   variables: Record<string, any>,
   options?: RequestOptions
 ): Promise<T> {
-  const response = await postForm(
-    url,
-    Object.assign(
-      { body: JSON.stringify({ query, variables }), headers: { 'Content-Type': 'application/json' } },
-      options
-    )
-  )
+  const response = (await postForm(url, {
+    body: JSON.stringify({ query, variables }),
+    headers: { 'Content-Type': 'application/json' },
+    ...options
+  })) as GraphQLResponse
   if (response.errors) {
     throw new Error(`Error querying graph. Reasons: ${JSON.stringify(response.errors)}`)
   }
   return response.data
 }
 
+async function identity<T>(a: T): Promise<T> {
+  return a
+}
+
 /**
  * This is the method where everything happens, all of the methods in this file call internally fetchInternal.
  * If you need to modify something for all requests, do it here.
  */
-async function fetchInternal<T>(
-  url: string,
-  responseConsumer: (response: Response) => Promise<T>,
-  options: CompleteRequestOptions
-): Promise<T> {
+async function fetchInternal(url: string, options: CompleteRequestOptions): Promise<Response> {
   return retry(
     async () => {
       const controller = new AbortController()
-      const timeout = setTimeout(() => {
-        controller.abort()
-      }, ms(options.timeout))
+      const transformRequest = options.requestMiddleware || identity
+      const transformResponse = options.responseMiddleware || identity
 
-      let request: CrossFetchRequest = {
+      const request = await transformRequest({
         requestInfo: url,
         requestInit: {
           signal: controller.signal,
           body: options.body,
           method: options.method,
           headers: getAllHeaders(options)
         }
-      }
-      if (options.requestMiddleware) {
-        request = await options.requestMiddleware(request)
-      }
+      })
+
+      // schedule timeout right after transforming Request
```

Comment:
Shouldn't we timeout the request middleware too? For example, when getting a new POW challenge?

Reply:
The timeouts are usually accounted only for network operations, if the middleware hangs that should be handled separately. It is easier to test in unit tests

Reply:
> @agusaldasoro bumping the major wasn't my intention, that's why I didn't removed any function and only added a `fetch` method. Should I do it anyways?

Okay, that's true, no need to bump major. My bad